fem/lininteg_domain.cpp

```diff
@@ -0,0 +1,339 @@
+               {
+                  const double zyX = QQQ(dz,dy,qx);
+                  for (int dx = 0; dx < d; ++dx) { u[dx] += zyX * Bt(dx,qx); }
+               }
+               for (int dx = 0; dx < d; ++dx) { Y(dx,dy,dz,c,e) += u[dx]; }
+            }
+         }
+         MFEM_SYNC_THREAD;
+      }
+   });
+}
+
+static void LaunchDeviceKernel(const FiniteElementSpace &fes,
+                               const IntegrationRule *ir,
+                               const Array<int> &markers,
+                               const Vector &coeff,
+                               Vector &y)
+{
+   Mesh *mesh = fes.GetMesh();
+   const int dim = mesh->Dimension();
+   const FiniteElement &el = *fes.GetFE(0);
+   const MemoryType mt = Device::GetDeviceMemoryType();
+   const DofToQuad &maps = el.GetDofToQuad(*ir, DofToQuad::TENSOR);
+   const int d = maps.ndof, q = maps.nqpt;
+   constexpr int flags = GeometricFactors::JACOBIANS;
+   const GeometricFactors *geom = mesh->GetGeometricFactors(*ir, flags, mt);
+   decltype(&DLFEvalAssemble2D<>) ker =
+      dim == 2 ? DLFEvalAssemble2D<> : DLFEvalAssemble3D<>;
+
+   if (dim==2)
+   {
+      if (d==2 && q==2) { ker=DLFEvalAssemble2D<2,2>; }
+      if (d==3 && q==3) { ker=DLFEvalAssemble2D<3,3>; }
+      if (d==4 && q==4) { ker=DLFEvalAssemble2D<4,4>; }
+      if (d==5 && q==5) { ker=DLFEvalAssemble2D<5,5>; }
+      if (d==2 && q==3) { ker=DLFEvalAssemble2D<2,3>; }
+      if (d==3 && q==4) { ker=DLFEvalAssemble2D<3,4>; }
+      if (d==4 && q==5) { ker=DLFEvalAssemble2D<4,5>; }
+      if (d==5 && q==6) { ker=DLFEvalAssemble2D<5,6>; }
+   }
+
+   if (dim==3)
+   {
+      if (d==2 && q==2) { ker=DLFEvalAssemble3D<2,2>; }
+      if (d==3 && q==3) { ker=DLFEvalAssemble3D<3,3>; }
+      if (d==4 && q==4) { ker=DLFEvalAssemble3D<4,4>; }
+      if (d==5 && q==5) { ker=DLFEvalAssemble3D<5,5>; }
+      if (d==2 && q==3) { ker=DLFEvalAssemble3D<2,3>; }
+      if (d==3 && q==4) { ker=DLFEvalAssemble3D<3,4>; }
+      if (d==4 && q==5) { ker=DLFEvalAssemble3D<4,5>; }
+      if (d==5 && q==6) { ker=DLFEvalAssemble3D<5,6>; }
+   }
+
+   MFEM_VERIFY(ker, "No kernel ndof " << d << " nqpt " << q);
+
+   const int vdim = fes.GetVDim();
+   const int ne = fes.GetMesh()->GetNE();
+   const int *M = markers.Read();
+   const double *B = maps.B.Read();
+   const double *J = geom->J.Read();
+   const double *W = ir->GetWeights().Read();
+   double *Y = y.ReadWrite();
+   ker(vdim, ne, d, q, M, B, J, W, coeff, Y);
+}
+
+void DomainLFIntegrator::AssembleDevice(const FiniteElementSpace &fes,
+                                        const Array<int> &markers,
+                                        Vector &b)
+{
+   const FiniteElement &fe = *fes.GetFE(0);
+   const int qorder = oa * fe.GetOrder() + ob;
+   const Geometry::Type gtype = fe.GetGeomType();
+   const IntegrationRule *ir = IntRule ? IntRule : &IntRules.Get(gtype, qorder);
+   const int nq = ir->GetNPoints(), ne = fes.GetMesh()->GetNE();
+
+   Vector coeff;
```

Comment:
I am looking forward to avoiding duplicating all of this logic once we get something like #2891!

Reply:
👍